meinberlin/apps/cache_helpers.py

```diff
@@ -0,0 +1,72 @@
+from logging import getLogger
+from typing import Optional
+
+from django.core.cache import cache
+from django.db.models import QuerySet
+from rest_framework.response import Response
+from rest_framework.viewsets import GenericViewSet
+
+from adhocracy4.organisations.models import Organisation
+
+log = getLogger(__name__)
+
+
+KEYS = {
+    "projects": set(),
```

Comment:
> you can't use (non-constant) globals as they won't be shared between the processes

Yes, but this is true for the the in-memory cache (`LocMemCache`) as well:
- https://docs.djangoproject.com/en/4.2/topics/cache/#local-memory-caching
- quote: `This cache is per-process (see below) and thread-safe.`

> with multiple threads and processes you can't use (non-constant) globals

As far as I know, our project is single threaded. In production we start it with gunicorn which in turn starts multiple workers (processes). The design of this cache is currently per-process, meaning each worker will have to cache a url (meaning up to "number of gunicorn workers" users may request a url before it is "fully" cached). I think in that case this setup works.

But the idea of using Redis sounds exciting:
- processes can share the cache
- we can glob cache keys (we don't have to keep track of them ourselves)

Reply:
If the cache is per-process, can we reliably clear the cache with signals ? I'm not super familiar with the signals and whether they are thread-safe (probably depending on your code for the signal) but they won't be process-safe, so deleting a cache won't reliable, right?

Reply:
> If the cache is per-process, can we reliably clear the cache with signals ?

Very good point! I think you are right. This can happen:
- 3 gunicorn workers are started
- all workers cache the projects
- worker 1 is called to delete a project
  - worker 1 triggers signal and clears its cache
- worker 2 and 3 still have projects cached because they did not delete the project / trigger the signal

Reply:
done, we are now only supporting redis if cache is enabled.